### PR TITLE
Incremental Feature Selection

### PR DESCRIPTION
**Pull request by Nama21yo**
## Description

Implemented `incremental feature selection` with configurable thresholds, redundancy checks, and adaptive binary search for optimal feature subset selection.

## Motivation and Context

Part of the effort to scale the MOSES MVP into supporting multiple deme expansion that depends on feature selection.

## How Has This Been Tested?

* All tests executed locally including edge-cases.

## Types of changes

* [ ] Bug fix (non-breaking change which fixes an issue)
* [x] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

* [x] My code follows the code style of this project.
* [x] I have added tests to cover my changes.
* [x] All new and existing tests passed.

### CODE REVIEW

**Review by Yagth — CHANGES_REQUESTED** [review]
Address the previous comments and it's good enough to merge.

File: feature-selection/feature-selection-helpers.metta [hunk 1]
```diff
@@ -274,4 +274,74 @@
...
+; Reference: https://github.com/opencog/cogutil/blob/88da482e38ba7ecec734dd0282d6c9b9be2c3bfd/opencog/util/algorithm.h#L282
+(: powerSet (-> Expression Number Bool Expression))
+(= (powerSet $set $n $exact)
+   (if (== $n 0)
+       (()) 
+       (let* (($ps   (powerSet $set (- $n 1) $exact))
+              ($new  (collapse 
+                       (let $subset (superpose $ps)
+                         (let $el (superpose $set)
+                           (if (== (isMember $el $subset) False)
```
[the remark this answers is not in the document]
Comment (Nama21yo) [reply feature-selection/feature-selection-helpers.metta:320]: Why I always do this kind of thing Yeah definitely will correct it

File: feature-selection/incremental.metta [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+                        (if (> $mutualInfo $threshold)
+                            ;; Add all indices from this relevant combination
+                            (chain (OS.union $resultSet $comboOS) $newResultSet
+                                (filterRelevantCombinations $itable $rest $threshold $newResultSet)
+                            )
+                            ;; Skip this combination
+                            (filterRelevantCombinations $itable $rest $threshold $resultSet)
+                        )
+                    )
+                )
```
Comment (K1rubel) [inline feature-selection/incremental.metta:61]: unify requires four argument, you forgot to add one here. 
```
(unify $1 ; the first pattern
       $2 ; the second pattern
       $3 ; the output for successful unification
       $4) ; fallback
```
you forgot to add the _fallback_.

File: feature-selection/incremental.metta [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+    (unify $subsetCombinations NilOS (Cons $currentBest (Cons $currentScore Nil))
+        (unify $subsetCombinations (ConsOS $subset $rest)
+            (chain (expressionToOS $subset NilOS) $subsetOS
+                (chain (calculateMutualInformation $itable $subsetOS) $score
+                    (if (> $score $currentScore)
+                        (findBestSubset $itable $rest $subset $score)
+                        (findBestSubset $itable $rest $currentBest $currentScore)
+                    )
+                )
+            )
```
Comment (K1rubel) [inline feature-selection/incremental.metta:83]: The same issue as above!

File: feature-selection/feature-selection-helpers.metta [hunk 4]
```diff
@@ -274,4 +274,74 @@
...
+
+;; Purpose: Check if difference between two numbers is within epsilon tolerance
+;; Parameters:
+;;   $min (Number) → Lower bound
+;;   $max (Number) → Upper bound  
+;;   $epsilon (Number) → Tolerance
+;; Returns: (Bool) → True if within epsilon
+;; Reference: https://github.com/opencog/cogutil/blob/6b149c457c9cfbf5edccefb166a7a9a1d2d74196/opencog/util/numeric.h#L178(: isWithin (-> Number Number Number Bool))
+(= (isWithin $min $max $epsilon)
+   (<= (abs-math (- $max $min)) $epsilon))
```
Comment (K1rubel) [inline feature-selection/feature-selection-helpers.metta:346]: I think this belongs in general helpers.

File: feature-selection/feature-selection-helpers.metta [hunk 5]
```diff
@@ -274,4 +274,74 @@
...
+;; Incremental Selection Utility Functions
+;; Purpose: Remove duplicate expressions from list using isExpEqual
+;; Recursively processes expressions and maintains unique accumulator
+;; Parameters:
+;;   $expr (Expression) → Expression containing potentially duplicate sub-expressions
+;;   $acc (Expression) → Accumulator for unique expressions
+;; Returns: (Expression) → Expression with duplicates removed
+;; Example: (featureSet ((A B) (B A) (C D)) ()) → ((A B) (C D))
+(: featureSet (-> Expression Expression Expression))
+(= (featureSet $expr $acc)
```
Comment (K1rubel) [inline feature-selection/feature-selection-helpers.metta:288]: Cant we find a better name for this one?

File: feature-selection/feature-selection-helpers.metta (hunk 4 above)
Comment (Nama21yo) [reply feature-selection/feature-selection-helpers.metta:346]: Would that be important to add it inside general-helpers I mean it is specifically related to feature selection.

File: utilities/ordered-set.metta [hunk 6]
```diff
@@ -82,4 +82,129 @@
...
+                    ;; Both expressions - compare first elements
+                      (let* (
+                           ($first1 (car-atom $atom1))
+                           ($first2 (car-atom $atom2))
+                           )
+                        (genericCompare $first1 $first2)
+                    )
+                    ;; atom1 is expression, atom2 is not - expressions come after atoms
+                    L)
+                ;; atom1 is not expression but atom2 is - atoms come before expressions
```
[the remark this answers is not in the document]
Comment (Nama21yo) [reply utilities/ordered-set.metta:114]: Yeah, you’re right. The purpose of genericCompare is just to handle duplicates like (A B) and (B A), so we don’t actually care about the exact ordering. That’s why I implemented it this way.

File: feature-selection/feature-selection-helpers.metta (hunk 4 above)
Comment (K1rubel) [reply feature-selection/feature-selection-helpers.metta:346]: I mean it is just a comparison operation, it is not sth that is specific to feature selection.

File: utilities/ordered-set.metta [hunk 7]
```diff
@@ -82,4 +82,129 @@
...
+
+;; Purpose: Union of two OrderedSets
+;; Parameters:
+;;   $set1 (OS Number) → First set
+;;   $set2 (OS Number) → Second set
+;; Returns: (OS Number) → Union of both sets
+;; Example: (OS.union (ConsOS 0 (ConsOS 1 NilOS)) (ConsOS 2 NilOS)) → (ConsOS 0 (ConsOS 1 (ConsOS 2 NilOS)))
+(: OS.union (-> (OS Number) (OS Number) (OS Number)))
+(= (OS.union $set1 $set2)
+    (unify $set1 NilOS $set2
```
[the remark this answers is not in the document]
Comment (Yagth) [reply utilities/ordered-set.metta]: But `case` is better than unify
